src/components/BaseComponents/Inputs/TextInput.tsx

```diff
@@ -44,31 +47,22 @@ const TextInput = ({
       // eslint-disable-next-line react/jsx-props-no-spreading
       {...rest}
     >
-      <InputGroup>
+      <InputGroup hasValidation>
         <Form.Control
           as="input"
-          onBlur={(e) =>
-            performValidation(
-              e,
-              validate,
-              setInvalidMessage,
-              setValidityChecked,
-            )
-          }
+          onBlur={validate}
           className={className}
           defaultValue={defaultValue}
           id={name}
           isInvalid={validityChecked && !!invalidMessage}
           isValid={validityChecked && !invalidMessage}
           name={name}
-          onChange={(e) => {
+          onChange={async (e) => {
             if (invalidMessage) {
-              performValidation(
-                e,
-                validate,
-                setInvalidMessage,
-                setValidityChecked,
-              );
+              const validationResult = validate(e);
+              if (validationResult instanceof Promise) {
```

Comment:
maybe could we explicitly type the validate function so that we can remove the lines where we do if (validationResult instanceof Promise) and instead just await validate()